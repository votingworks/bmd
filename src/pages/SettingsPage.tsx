import React, { PointerEventHandler, useContext } from 'react'
import styled from 'styled-components'
import GLOBALS from '../config/globals'

import { InputEvent, TextSizeSetting } from '../config/types'

import BallotContext from '../contexts/ballotContext'

import Button from '../components/Button'
import ButtonBar from '../components/ButtonBar'
import LinkButton from '../components/LinkButton'
import Main, { MainChild } from '../components/Main'
import Prose from '../components/Prose'
import RangeInput from '../components/RangeInput'

const SettingLabel = styled.label`
  display: block;
  margin: 1.5rem 0 1rem;
  font-size: 1.25rem;
  font-weight: 700;
`
const FontSizeControlsContainer = styled.div`
  display: flex;
  & input {
    flex: 1;
    margin-right: 0.5rem;
    margin-left: 0.5rem;
  }
  button {
    border-radius: 50%;
    width: 2.5rem;
    height: 2.5rem;
    padding: 0 0 0.2rem;
  }
`

const SettingsPage = () => {
  const { userSettings, setUserSettings } = useContext(BallotContext)
  const onFontSizeChange = (event: InputEvent) => {
    const target = event.target as HTMLInputElement
    const textSize = +target.value as TextSizeSetting
    setUserSettings({ textSize })
  }
  const adjustFontSize: PointerEventHandler = event => {
    const target = event.target as HTMLButtonElement
    const textSize = (userSettings.textSize + +target.value) as TextSizeSetting
    setUserSettings({ textSize })
  }
  return (
    <>
      <Main>
        <MainChild>
          <Prose>
            <h1>Settings</h1>
            <p>Adjust the following settings to meet your needs.</p>
            <SettingLabel htmlFor="font-size">Font Size</SettingLabel>
            <FontSizeControlsContainer>
              <Button
                aria-hidden
                data-testid="decrease-font-size-button"
                disabled={userSettings.textSize === 0}
                onPress={adjustFontSize}
                value={-1}
              >
                -
              </Button>
              <RangeInput
                id="font-size"
                min={0}
                max={3}
                step={1}
                value={userSettings.textSize}
                onChange={onFontSizeChange}
              />
              <Button
                aria-hidden
                data-testid="increase-font-size-button"
                disabled={
                  userSettings.textSize === GLOBALS.FONT_SIZES.length - 1
                }
                onPress={adjustFontSize}
                value={1}
              >
                +
              </Button>
            </FontSizeControlsContainer>
          </Prose>
        </MainChild>
      </Main>
      <ButtonBar>
        <div />
        <LinkButton goBack>Back</LinkButton>
        <div />
        <div />
      </ButtonBar>
    </>
  )
}

export default SettingsPage
