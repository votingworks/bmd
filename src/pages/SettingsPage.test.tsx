import { axe } from 'jest-axe'
import React from 'react'
import { Route } from 'react-router-dom'

import { render } from '../../test/testUtils'

import SettingsPage from './SettingsPage'

it(`renders SettingsPage`, async () => {
  const { container } = render(<Route path="/" component={SettingsPage} />, {
    route: '/',
  })
  expect(container.firstChild).toMatchSnapshot()
  expect(await axe(container.innerHTML)).toHaveNoViolations()
})
