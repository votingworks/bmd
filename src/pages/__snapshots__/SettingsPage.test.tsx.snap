// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SettingsPage 1`] = `
.c5 {
  border: none;
  border-radius: 0.25rem;
  box-sizing: border-box;
  background: rgb(211,211,211);
  cursor: pointer;
  padding: 0.75rem 1rem;
  line-height: 1.25;
  color: black;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: auto;
  padding: 1rem 0.5rem 2rem;
}

.c1 {
  margin: 0 0;
  max-width: 35rem;
}

.c2 {
  width: 100%;
  max-width: 66ch;
  line-height: 1.2;
}

.c2 h1 {
  margin: 2rem 0 1rem;
  line-height: 1.1;
  font-size: 1.5rem;
}

.c2 h2 {
  margin: 1.5rem 0 0.75rem;
  font-size: 1.25rem;
}

.c2 h3,
.c2 p {
  margin-top: 1rem;
  margin-bottom: 1rem;
  font-size: 1rem;
}

.c2 h1 + h2 {
  margin-top: -0.75rem;
}

.c2 h1 + p,
.c2 h2 + p {
  margin-top: -0.75rem;
}

.c2 h3 + p {
  margin-top: -1rem;
}

.c2 > :first-child {
  margin-top: 0;
}

.c2 > :last-child {
  margin-bottom: 0;
}

.c6 {
  background: transparent;
  width: 300px;
}

.c6,
.c6::-webkit-slider-runnable-track,
.c6::-webkit-slider-thumb {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c6:focus {
  outline: none;
}

.c6::-webkit-slider-runnable-track {
  cursor: pointer;
  width: 100%;
  height: 0.5rem;
  margin: 0.25rem 0;
  background: radial-gradient( circle 0.25rem at calc(0% + 0.5rem) 0.25rem, black, black 100%, transparent ), radial-gradient( circle 0.25rem at calc((100% + 0.125rem) * .34) 0.25rem, black, black 100%, transparent ), radial-gradient( circle 0.25rem at calc((100% - 0.125rem) * .66) 0.25rem, black, black 100%, transparent ), radial-gradient( circle 0.25rem at calc(100% - 0.5rem) 0.25rem, black, black 100%, transparent ), linear-gradient( to right, transparent, transparent 0.5rem, rgb(211,211,211) 0.5rem, rgb(211,211,211) calc(100% - 0.5rem), transparent calc(100% - 0.5rem), transparent ),#edeff0;
  border-color: transparent;
  color: transparent;
}

.c6:focus::-webkit-slider-runnable-track {
  outline: none;
}

.c6::-webkit-slider-thumb {
  height: 1rem;
  width: 1rem;
  border-radius: 0.5rem;
  background: black;
  cursor: pointer;
  margin-top: -0.25rem;
}

.c6:focus::-webkit-slider-thumb {
  outline: rgb(77,144,254) dashed 0.25rem;
}

.c3 {
  display: block;
  margin: 1.5rem 0 1rem;
  font-size: 1.25rem;
  font-weight: 700;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 input {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin-right: 0.5rem;
  margin-left: 0.5rem;
}

.c4 button {
  border-radius: 50%;
  width: 2.5rem;
  height: 2.5rem;
  padding: 0 0 0.2rem;
}

@media (min-width:480px) {
  .c0 {
    padding: 1rem 1rem 2rem;
  }
}

@media print {
  .c0 {
    -webkit-box-pack: start;
    -webkit-justify-content: flex-start;
    -ms-flex-pack: start;
    justify-content: flex-start;
    padding: 0;
  }
}

@media print {
  .c1 {
    margin: 0;
    max-width: 100%;
  }
}

@media (min-width:480px) {
  .c2 {
    line-height: 1.3;
  }
}

<main
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <h1>
        Settings
      </h1>
      <p>
        Adjust the following settings to meet your needs.
      </p>
      <label
        class="c3"
        for="font-size"
      >
        Font Size
      </label>
      <div
        class="c4"
      >
        <button
          aria-hidden="true"
          class="c5"
          data-testid="decrease-font-size-button"
          type="button"
          value="-1"
        >
          -
        </button>
        <input
          class="c6"
          id="font-size"
          max="3"
          min="0"
          step="1"
          type="range"
          value="1"
        />
        <button
          aria-hidden="true"
          class="c5"
          data-testid="increase-font-size-button"
          type="button"
          value="1"
        >
          +
        </button>
      </div>
      <h2>
        Clear Selections
      </h2>
      <p>
        Clear all selections and start over.
      </p>
      <button
        class="c5"
        type="button"
      >
        Start Over
      </button>
    </div>
  </div>
</main>
`;
